### PR TITLE
Compiler Refactor

### PR DESCRIPTION
**Pull request by Dashadower**
Sorry for another ginormous commit. This commit has PR #56 merged in, so merging this will also merge said PR.

Features:
- refactored `rat.compiler.compiler()` into `rat.compiler.Compiler()` class that breaks down the previously big function into smaller ones. You can now access the compile results through `Compiler(data_df, expr_tree, model_code).compile())`
- Topological ordering of expression tress as to just variables. Previously we had to run a nested for loop with the outer along target variables and inner along code lines/trees. Now the trees themselves are ordered, which allows just a single loop and no possibility of N! worst case running time.
- Better compile error messages and detection. Compiler also has detailed error messages much like the Parser and Scanner. In addition, a semi-full rewrite resulted in less dangling edge cases, so a standard python exception being raised when there's something wrong with the code wouldn't happen frequently.
```
rat.compiler.CompileError: An error occured while compiling the following line(2:21):
theta[school] = mu + z[schools] * tau;
                     ^
Couldn't index RHS variable z's declared subscript('schools',) from LHS base df, which has subscripts ('school',).
```
```
rat.compiler.CompileError: An error occured while compiling the following line(1:49):
score_diff ~normal (skills[ home_team , year ] - skillss[away_team, year],sigma[year]);
                                                 ^
Could not find variable skillss
```
```
rat.compiler.CompileError: An error occured while compiling the following line(2:0):
skills[team, year] ~ normal(skills_mu[year], 1);
^
Could not topologically order the expression tree. This is because your model is cyclic.
```

### CODE REVIEW

**Comment by Dashadower** [comment]
This PR fixes #41, #43, #44

**Comment by bbbales2** [comment]
Yo this looks sick.

How does this error work though:
```
rat.compiler.CompileError: An error occured while compiling the following line(1:49):
score_diff ~normal (skills[ home_team , year ] - skillss[away_team, year],sigma[year]);
                                                 ^
Could not find variable skillss
```

In my mind not finding a variable isn't an error -- the compiler should just infer one exists cause the use.

File: rat/ops.py [hunk 1]
```diff
@@ -21,6 +24,8 @@ def __str__(self):
 @dataclass(frozen=True)
 class RealConstant(Expr):
     value: float
+    line_index: int = -1  # line index of the original model code
+    column_index: int = -1  # column index of the original model code
```
Comment (bbbales2) [inline rat/ops.py:28]: I think we can put the defaults in the base class (and make Expr a dataclass). There's a trick to make this work, but I think it should be possible: https://stackoverflow.com/a/58525728.

**Comment by Dashadower** [comment]
> Yo this looks sick.
> 
> How does this error work though:
> 
> ```
> rat.compiler.CompileError: An error occured while compiling the following line(1:49):
> score_diff ~normal (skills[ home_team , year ] - skillss[away_team, year],sigma[year]);
>                                                  ^
> Could not find variable skillss
> ```
> 
> In my mind not finding a variable isn't an error -- the compiler should just infer one exists cause the use.

Well, if you mean it shouldn't raise an error because not defining a prior(LHS use) for a variable doesn't mean it doesn't exist, then yes it should be changed to add a default prior like Uniform(-2, 2). I think this is what you meant.

But currently the assumption is that in order for a variable to exist it must be used at least once on the LHS. Since the lines are topologically ordered if we haven't seen `skillss` that means it wasn't used on LHS and thus is a non-existant variable.

**Comment by Dashadower** [comment]
What should be the default prior be? Uniform(-2, 2) like stan?

**Comment by bbbales2** [comment]
> Well, if you mean it shouldn't raise an error because not defining a prior(LHS use) for a variable doesn't mean it doesn't exist

Oh oh, I see, so like a prior hasn't been defined for this. I can live with priors being required for now. It's not obvious to me that we don't want something like this eventually.

In this case could the error be something like `No prior found for skillss`?

**Comment by Dashadower** [comment]
> > Well, if you mean it shouldn't raise an error because not defining a prior(LHS use) for a variable doesn't mean it doesn't exist
> 
> Oh oh, I see, so like a prior hasn't been defined for this. I can live with priors being required for now. It's not obvious to me that we don't want something like this eventually.
> 
> In this case could the error be something like `No prior found for skillss`?

Yeah I think that's a better error message.

But it wouldn't be too hard to throw in a default prior in the current code. [This line](https://github.com/bbbales2/regressions/blob/49253f6b0dd6a113f8faeff79fe1e3dd51569f4c/rat/compiler.py#L506) is the only thing that we have to change; just create a `variable.Param` and a simple linefunction with the default prior.

**Comment by bbbales2** [comment]
I don't think default priors for now. For now priors required is good, and in the future we might want to consider improper priors (just don't do anything), and I haven't really thought so much about actual default priors in the way that rstanarm has them. Even if we had default from data priors, I might want to make them like:

```
mu ~ rstanarm_default();
```

To at least have an indicator that something is happening.

**Review by bbbales2 — APPROVED** [review]
This all looks very nice. Good cleanup. I'm gonna try to use the default values for the Expr class. If that works, I'll add it and merge this all. If it doesn't, I'll just merge this as-is.

File: rat/ops.py (hunk 1 above)
Comment (Dashadower) [reply rat/ops.py:28]: Sounds good. I was quite frustrated having to copy-paste those two lines repeatedly.

Comment (bbbales2) [reply rat/ops.py]: Cool beans. It's some Python 3.10 stuff that makes this easy: https://stackoverflow.com/a/69822584

**Pull request merged**